airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteValueDeepCoercingMapper.kt

```diff
@@ -0,0 +1,231 @@
+
+    override fun mapNumber(value: AirbyteValue, context: Context): Pair<AirbyteValue, Context> =
+        when (value) {
+            is NumberValue -> value to context
+            is IntegerValue -> NumberValue(value.value.toBigDecimal()) to context
+            is StringValue -> NumberValue(BigDecimal(value.value)) to context
+            else -> nulledOut(NumberType, context)
+        }
+
+    override fun mapInteger(value: AirbyteValue, context: Context): Pair<AirbyteValue, Context> =
+        when (value) {
+            // Maybe we should truncate non-int values?
+            // But to match existing behavior, let's just null for now.
+            is NumberValue -> IntegerValue(value.value.toBigIntegerExact()) to context
+            is IntegerValue -> value to context
+            is StringValue -> IntegerValue(BigInteger(value.value)) to context
+            else -> nulledOut(IntegerType, context)
+        }
+
+    override fun mapString(value: AirbyteValue, context: Context): Pair<AirbyteValue, Context> {
+        val stringified =
+            when (value) {
+                // this should never happen, because we handle `value is NullValue`
+                // in the top-level if statement
+                NullValue -> throw IllegalStateException("Unexpected NullValue")
+                is StringValue -> value.value
+                is NumberValue -> value.value.toString()
+                is IntegerValue -> value.value.toString()
+                is BooleanValue -> value.value.toString()
+                is ArrayValue,
+                is ObjectValue -> value.serializeToString()
+                // JsonToAirbyteValue never outputs these values, so don't handle them.
+                is DateValue,
+                is TimeWithTimezoneValue,
+                is TimeWithoutTimezoneValue,
+                is TimestampWithTimezoneValue,
+                is TimestampWithoutTimezoneValue,
+                is UnknownValue ->
+                    throw IllegalArgumentException(
+                        "Invalid value type ${value.javaClass.canonicalName}"
+                    )
+            }
+        return StringValue(stringified) to context
+    }
+
+    override fun mapDate(value: AirbyteValue, context: Context): Pair<AirbyteValue, Context> =
+        requireType<StringValue>(value, DateType, context) {
+            DateValue(LocalDate.parse(it.value, DATE_TIME_FORMATTER)) to context
+        }
+
+    override fun mapTimeWithTimezone(
+        value: AirbyteValue,
+        context: Context
+    ): Pair<AirbyteValue, Context> =
+        requireType<StringValue>(value, TimeTypeWithTimezone, context) {
+            val ot =
+                try {
+                    OffsetTime.parse(it.value, TIME_FORMATTER)
+                } catch (e: Exception) {
+                    LocalTime.parse(it.value, TIME_FORMATTER).atOffset(ZoneOffset.UTC)
+                }
+            TimeWithTimezoneValue(ot) to context
+        }
```

Comment:
except that this requires that I be a string coming in. So i would have to put `ToTimestamps` (and `Schemaless`) *after* this. 

```Marshal -> [Pipeline: MergeUnions -> Coerce -> MapTimestamps -> SchemalessToStrings -> 
 ToDisjointRecord] -> ToAvro```

Hm... having the extra types would make this more flexible, yes? B/c they would be validated here. then i could do `MapTimestamps -> SchemalessToStrings -> MergeUnions -> ToDisjointRecord -> Coerce(ie validate)`, which makes more sense. Like, first map the types, then resolve the unions, then change how unions are represented, finally validate that all the values match their final types, then convert?

Reply:
hm. maybe a thing we should pick up once we're back from holidays, I think I'm mostly with you, but confused about putting MapTimestamps first. I'm thinking of Coerce as The Thing That Maps to Real Types, so anything that wants to operate on timestamps should come after Coerce

(but yeah, putting the schemalesstostring + todisjointrecord before Coerce is pretty clearly superior)